pkcs8/src/private_key_info.rs

```diff
@@ -250,45 +290,127 @@ impl<'a> fmt::Debug for PrivateKeyInfo<'a> {
+impl<'a, Params, Key, PubKey> TryFrom<PrivateKeyInfoInner<Params, Key, PubKey>> for SecretDocument
+where
+    Params: der::Choice<'a, Error = der::Error> + Encode,
+    Key: DecodeValue<'a, Error = der::Error> + FixedTag + 'a,
+    Key: EncodeValue,
+    PubKey: DecodeValue<'a, Error = der::Error> + FixedTag + 'a,
+    PubKey: BitStringLike,
+{
     type Error = Error;
 
-    fn try_from(private_key: PrivateKeyInfo<'_>) -> Result<SecretDocument> {
+    fn try_from(private_key: PrivateKeyInfoInner<Params, Key, PubKey>) -> Result<SecretDocument> {
         SecretDocument::try_from(&private_key)
     }
 }
 
 #[cfg(feature = "alloc")]
-impl TryFrom<&PrivateKeyInfo<'_>> for SecretDocument {
+impl<'a, Params, Key, PubKey> TryFrom<&PrivateKeyInfoInner<Params, Key, PubKey>> for SecretDocument
+where
+    Params: der::Choice<'a, Error = der::Error> + Encode,
+    Key: DecodeValue<'a, Error = der::Error> + FixedTag + 'a,
+    Key: EncodeValue,
+    PubKey: DecodeValue<'a, Error = der::Error> + FixedTag + 'a,
+    PubKey: BitStringLike,
+{
     type Error = Error;
 
-    fn try_from(private_key: &PrivateKeyInfo<'_>) -> Result<SecretDocument> {
+    fn try_from(private_key: &PrivateKeyInfoInner<Params, Key, PubKey>) -> Result<SecretDocument> {
         Ok(Self::encode_msg(private_key)?)
     }
 }
 
 #[cfg(feature = "pem")]
-impl PemLabel for PrivateKeyInfo<'_> {
+impl<Params, Key, PubKey> PemLabel for PrivateKeyInfoInner<Params, Key, PubKey> {
     const PEM_LABEL: &'static str = "PRIVATE KEY";
 }
 
 #[cfg(feature = "subtle")]
-impl<'a> ConstantTimeEq for PrivateKeyInfo<'a> {
+impl<Params, Key, PubKey> ConstantTimeEq for PrivateKeyInfoInner<Params, Key, PubKey>
+where
+    Params: Eq,
+    Key: PartialEq + AsRef<[u8]>,
+    PubKey: PartialEq,
+{
     fn ct_eq(&self, other: &Self) -> Choice {
         // NOTE: public fields are not compared in constant time
         let public_fields_eq =
             self.algorithm == other.algorithm && self.public_key == other.public_key;
 
-        self.private_key.ct_eq(other.private_key) & Choice::from(public_fields_eq as u8)
+        self.private_key.as_ref().ct_eq(other.private_key.as_ref())
+            & Choice::from(public_fields_eq as u8)
     }
 }
 
 #[cfg(feature = "subtle")]
-impl<'a> Eq for PrivateKeyInfo<'a> {}
+impl<Params, Key, PubKey> Eq for PrivateKeyInfoInner<Params, Key, PubKey>
+where
+    Params: Eq,
+    Key: AsRef<[u8]> + Eq,
+    PubKey: Eq,
+{
+}
 
 #[cfg(feature = "subtle")]
-impl<'a> PartialEq for PrivateKeyInfo<'a> {
+impl<Params, Key, PubKey> PartialEq for PrivateKeyInfoInner<Params, Key, PubKey>
+where
+    Params: Eq,
+    Key: PartialEq + AsRef<[u8]>,
+    PubKey: PartialEq,
+{
     fn eq(&self, other: &Self) -> bool {
         self.ct_eq(other).into()
     }
 }
+
+/// [`PrivateKeyInfoInner`] with [`AnyRef`] algorithm parameters, and `&[u8]` key.
+pub type PrivateKeyInfo<'a> = PrivateKeyInfoInner<AnyRef<'a>, OctetStringRef<'a>, BitStringRef<'a>>;
```

Comment:
Shouldn't this be `PrivateKeyInfoRef` and the generic struct retained as `PrivateKeyInfo`, following what we did in `spki`?

Reply:
Yeah, I thought it would be less of an API break, but I'm breaking the API anyway.